x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java

```diff
@@ -233,10 +238,23 @@ public Settings additionalSettings() {
                                                IndexNameExpressionResolver expressionResolver) {
         List<Object> components = new ArrayList<>();
 
-        final SSLService sslService = new SSLService(environment);
+        final Map<String, SSLConfiguration> sslConfigurations = SSLService.getSSLConfigurations(environment.settings());
+        final CompletableFuture<SSLService> sslServiceFuture = new CompletableFuture<>();
+        final Consumer<SSLConfiguration> reloadConsumer = sslConfiguration -> {
+            try {
+                final SSLService sslService = sslServiceFuture.get();
+                logger.debug("reloading ssl configuration [{}]", sslConfiguration);
+                sslService.reloadSSLContext(sslConfiguration);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (ExecutionException e) {
+                throw new ElasticsearchException("failed to obtain ssl service", e);
+            }
+        };
```

Comment:
Can we encapsulate this logic inside `SSLConfigurationReloader` like how it was before? I personally prefer to have `SSLConfigurationReloader` hide the complexitiy as much as possible. How about have APIs like the follows:
* `sslConfigurationReloader = new SSLConfigurationReloader(...)` to prepare everything and also create the Future object internally. So everything is ready just waiting for the future to be completed.
* `sslConfigurationReloader.startWatching(sslService)` to resolve the future internally. Effectively this method is just oneline of `sslServiceFuture.complete(sslService)`

Reply:
I'd rather not put this logic in SSLConfigurationReloader for a few reasons.

1. The need for a future is tied directly to the order of initialization and that should not be something that the configuration reloader needs to be concerned with.
2. We avoid creating an object to keep around for the lifetime of the node by leaving this initialization logic out of the SSLConfigurationReloader class
3. The SSLConfigurationReloader and SSLService are not directly tied together; our reload change listeners just need to know that some consumer needs to be called if a file changes

My alternative is to "hide" the start of reloading and the glue between the two in a private method to create the SSLService, which I've done in https://github.com/elastic/elasticsearch/pull/54999/commits/6830389674428336d0ef21883d1e4b04c3e0c1e3

Reply:
I personally still think the separation between the reloadConsumer setup and the reloader itself is unnecessary. 

It has value in theory and also the point of "one less object to keep around" is good. But in the current form (both before and after this PR), `SSLConfigurationReloader` is specific about `SSLService`. It does one specific thing which is to ensure a call to `sslService.reloadSSLContext(...)` on file changes. It can be generalised by keep differnt parts in separate classes. But in fact we don't have a 2nd use case for it. So this decomposition feels unnecessary to me. 

In my opinion, having these logics encapsulated by `SSLConfigurationReloader` makes the code easier to understand and clean to use. The fact that "a Reloader should guarantee an update is always picked up" sounds like internal implementation details to me, regardless whether it uses a `Future` or something else to achieve it. So I'd personally leave that to the Reloader class itself.

Reply:
I've gone ahead with a version that moves this back into the reloader.

Reply:
Thanks @jaymode I appreciate the open discussion.